nixos/modules/services/torrent/deluge.nix

```diff
@@ -20,45 +44,143 @@ in {
+          '';
+        };
+
+        authFile = mkOption {
+          type = types.path;
+          example = "/run/keys/deluge-auth";
+          description = ''
+            The file managing the authentication for deluge, the format of
+            this file is straightforward, each line contains a
+            username:password:level tuple in plaintext.
+            Only has an effect when <option>services.deluge.declarative</option>
+            is set to <literal>true</literal>.
+          '';
+        };
       };
 
-      deluge.web.enable = mkEnableOption "Deluge Web daemon";
+      deluge.web = {
+        enable = mkEnableOption "Deluge Web daemon";
+
+        port = mkOption {
+        type = types.ints.u16;
+          example = 8112;
+          description = ''
+            Deluge web UI port.
+          '';
+        };
+      };
     };
   };
 
   config = mkIf cfg.enable {
 
+    systemd.tmpfiles.rules = [ "d '${configDir}' 0700 deluge deluge" ]
+    ++ optional (hasAttr "download_location" cfg.config) "d '${cfg.config.download_location}' 0770 deluge deluge"
+    ++ optional (hasAttr "torrentfiles_location" cfg.config) "d '${cfg.config.torrentfiles_location}' 0770 deluge deluge"
+    ++ optional (hasAttr "move_completed_path" cfg.config) "d '${cfg.config.move_completed_path}' 0770 deluge deluge";
+
     systemd.services.deluged = {
       after = [ "network.target" ];
       description = "Deluge BitTorrent Daemon";
       wantedBy = [ "multi-user.target" ];
       path = [ pkgs.deluge ];
       serviceConfig = {
-        ExecStart = "${pkgs.deluge}/bin/deluged -d";
+        ExecStart = ''
+          ${pkgs.deluge}/bin/deluged \
+            --do-not-daemonize \
+            --config ${configDir}
+        '';
         # To prevent "Quit & shutdown daemon" from working; we want systemd to manage it!
         Restart = "on-success";
         User = "deluge";
         Group = "deluge";
+        UMask = "0002";
         LimitNOFILE = cfg.openFilesLimit;
       };
+      preStart = preStart;
     };
 
     systemd.services.delugeweb = mkIf cfg_web.enable {
-      after = [ "network.target" ];
+      after = [ "network.target" "deluged.service"];
+      requires = [ "deluged.service" ];
       description = "Deluge BitTorrent WebUI";
       wantedBy = [ "multi-user.target" ];
       path = [ pkgs.deluge ];
-      serviceConfig.ExecStart = "${pkgs.deluge}/bin/deluge --ui web";
-      serviceConfig.User = "deluge";
-      serviceConfig.Group = "deluge";
+      serviceConfig = {
+        ExecStart = ''
+          ${pkgs.deluge}/bin/deluge-web \
+            --config ${configDir} \
+            ${optionalString (cfg.web.port != null) "--port ${toString cfg.web.port}"}
+        '';
+        User = "deluge";
+        Group = "deluge";
+      };
+    };
+
+    networking.firewall = mkIf cfg.openFirewall {
+      allowedTCPPorts = optional (hasAttr "daemon_port" cfg.config) cfg.config.daemon_port
+        ++ optional cfg_web.enable cfg.web.port;
```

Comment:
Oh and another thing I just noticed: Deluge has a fixed initial password for the web interface. So if you enable it and the port gets opened, anybody could just access the web client from outside, change the password, and screw stuff up. We can't leave it like this.

So you should either somehow make the password declarative so users have to specify it in their config, or add an `openFirewall` option specifically for the deluge-web, disabled by default, so people can set their password first locally, and then open the ports.

Reply:
Good catch, that's one stupid security hole I didn't thought trough...

Even if the web interface port isn't open when the default password is in use the password can be MiTM since I didn't added the possibility to configure the SSL files. With that in mind and the fact
that web UI can be used to connect to a remote server it doesn't make sense to open it's port ever.
I think `deluge-wen` should only be seen as any other deluge client interface like `deluge-console` and `deluge-gtk`.

Reply:
Thinking about it some more the `daemon_port` too shouldn't be open unless some SSL certificates are in place. The [doc](https://dev.deluge-torrent.org/wiki/UserGuide/ThinClient#CreateSSHTunnel) explain how to connect to the daemon securely.
WDYT?

Reply:
That's a bit of a bummer.. If we don't open any ports at all, we should mention in the `enable`'s option description how to connect from outside. Linking to those docs on SSH forwarding would be good, and mention VPN as well. In addition we should mention the security risks of opening the ports.